src/main/java/com/cosmian/jna/cover_crypt/Ffi.java

```diff
@@ -658,10 +663,139 @@ public static byte[] decryptBlock(byte[] symmetricKey, byte[] uid, int blockNumb
+        }
+
+        int privateKeySize = ByteBuffer.wrap(Arrays.copyOfRange(masterKeysBytes, 0, 4)).getInt();
+        byte[] privateKey = Arrays.copyOfRange(masterKeysBytes, 4, 4 + privateKeySize);
+        byte[] publicKey = Arrays.copyOfRange(masterKeysBytes, 4 + privateKeySize, masterKeysBufferSize.getValue());
+
+        return new MasterKeys(privateKey, publicKey);
+    }
+
+    /**
+     * Generate the user private key
+     *
+     * @param masterPrivateKey the master private key in bytes
+     * @param accessPolicy the access policy of the user private key
+     * @param policy the CoverCrypt policy
+     * @return the corresponding user private key
+     * @throws FfiException in case of native library error
+     */
+    public static byte[] generateUserPrivateKey(byte[] masterPrivateKey, AccessPolicy accessPolicy, Policy policy)
+        throws FfiException {
+        // User private key Bytes OUT
+        byte[] userPrivateKeyBuffer = new byte[8192];
+        IntByReference userPrivateKeyBufferSize = new IntByReference(userPrivateKeyBuffer.length);
+
+        // Master private key
+        final Pointer masterPrivateKeyPointer = new Memory(masterPrivateKey.length);
+        masterPrivateKeyPointer.write(0, masterPrivateKey, 0, masterPrivateKey.length);
+
+        // For the JSON strings
+        ObjectMapper mapper = new ObjectMapper();
+
+        // Access Policy
+        String accessPolicyJson;
+        try {
+            accessPolicyJson = mapper.writeValueAsString(accessPolicy);
+        } catch (JsonProcessingException e) {
+            throw new FfiException("Invalid Access Policy");
+        }
+        // Policy
+        String policyJson;
+        try {
+            policyJson = mapper.writeValueAsString(policy);
+        } catch (JsonProcessingException e) {
+            throw new FfiException("Invalid Policy");
+        }
+
+        unwrap(Ffi.INSTANCE.h_generate_user_private_key(userPrivateKeyBuffer, userPrivateKeyBufferSize,
+            masterPrivateKeyPointer, masterPrivateKey.length, accessPolicyJson, policyJson));
+
+        return Arrays.copyOfRange(userPrivateKeyBuffer, 0, userPrivateKeyBufferSize.getValue());
+    }
+
+    /**
+     * Rotate attributes, changing their underlying value with that of an unused slot
+     *
+     * @param attributes: a list of attributes to rotate
+     * @param policy: the current policy returns the new Policy
+     * @return the new policy
+     * @throws FfiException in case of native library error
+     * @throws IOException standard IO exceptions
+     * @throws DatabindException standard databind exceptions
+     * @throws StreamReadException stream read exceptions
+     */
+    public static Policy rotateAttributes(Attr[] attributes, Policy policy) throws FfiException, StreamReadException, DatabindException, IOException {
+        // New policy Bytes OUT
+        byte[] policyBuffer = new byte[4096];
```

Comment:
This buffer looks small; I think there should be a retry mechanism in case the buffer is too small:
- FFI call returns
  - 0 if success
  - -1 if unrecoverable failure
  - any positive umber: increase buffer size to that number and retry